logprep/processor/calculator/fourFn.py

```diff
@@ -0,0 +1,163 @@
+# pylint: disable=missing-docstring
+# pylint: disable=invalid-name
+"""
+based on https://github.com/pyparsing/pyparsing/blob/master/examples/fourFn.py
+
+Demonstration of the pyparsing module, implementing a simple 4-function expression parser,
+with support for scientific notation, and symbols for e and pi.
+Extended to add exponentiation and simple built-in functions.
+Extended test cases, simplified pushFirst method.
+Removed unnecessary expr.suppress() call (thanks Nathaniel Peterson!), and added Group
+Changed fnumber to use a Regex, which is now the preferred method
+Reformatted to latest pypyparsing features, support multiple and variable args to functions
+Copyright 2003-2019 by Paul McGuire
+"""
+
+import math
+import operator
+
+from pyparsing import (
+    CaselessKeyword,
+    Forward,
+    Group,
+    Literal,
+    Regex,
+    Suppress,
+    Word,
+    alphanums,
+    alphas,
+    delimitedList,
+)
+
+exprStack = []
+
+
+def push_first(toks):
+    exprStack.append(toks[0])
+
+
+def push_unary_minus(toks):
+    for t in toks:
+        if t == "-":
+            exprStack.append("unary -")
+        else:
+            break
+
+
+bnf = None
+
+
+def BNF():
+    """
+    expop   :: '^'
+    multop  :: '*' | '/'
+    addop   :: '+' | '-'
+    integer :: ['+' | '-'] '0'..'9'+
+    atom    :: PI | E | real | fn '(' expr ')' | '(' expr ')'
+    factor  :: atom [ expop factor ]*
+    term    :: factor [ multop factor ]*
+    expr    :: term [ addop term ]*
+    """
+    global bnf  # pylint: disable=global-statement
```

Comment:
If we wrap everything in a separate class we can maybe also get rid of this `global` statement.